dwave/optimization/include/dwave-optimization/nodes/mathematical.hpp

```diff
@@ -151,6 +151,73 @@ using NaryMaximumNode = NaryOpNode<functional::max<double>>;
 using NaryMinimumNode = NaryOpNode<functional::min<double>>;
 using NaryMultiplyNode = NaryOpNode<std::multiplies<double>>;
 
+/// TODO: support multiple axes
+template <class BinaryOp>
+class PartialReduceNode : public ArrayOutputMixin<ArrayNode> {
+ public:
+    // Runtime constructor
+    PartialReduceNode(ArrayNode* array_ptr, ssize_t axis, double init);
+
+    // Some operations have known default values so we can create them regardless of
+    // whether or not the array is dynamic.
+    // Others will raise an error for non-dynamic arrays.
+    explicit PartialReduceNode(ArrayNode* node_ptr, ssize_t axis);
+
+    double const* buff(const State& state) const override;
+    std::span<const Update> diff(const State& state) const override;
+    bool integral() const override;
+    double max() const override;
+    double min() const override;
+
+    using ArrayOutputMixin::shape;
+    std::span<const ssize_t> shape(const State& state) const override;
+
+    using ArrayOutputMixin::size;
+    ssize_t size(const State& state) const override;
+
+    ssize_t size_diff(const State& state) const override;
+
+    void commit(State& state) const override;
+    void revert(State& state) const override;
+    void initialize_state(State& state) const override;
+    void propagate(State& state) const override;
+
+    // The predecessor of the reduction, as an Array*.
+    std::span<Array* const> operands() {
+        assert(predecessors().size() == 1);
+        return std::span<Array* const, 1>(&array_ptr_, 1);
+    }
+    std::span<const Array* const> operands() const {
+        assert(predecessors().size() == 1);
+        return std::span<const Array* const, 1>(&array_ptr_, 1);
+    }
+
+    const std::optional<double> init;
+
+    ssize_t axis() const { return axis_; }
```

Comment:
If we think there is a decent chance we want to support multiple axis in the future, then IMO
```c++

    std::span<const ssize_t> axis() const { ... }
```
will be more future proof. And we should also move the implementation to the `.cpp` for binary compatibility.

Reply:
Or should it be `axes`?

Reply:
> Or should it be axes?
For consistency with numpy we can leave it as `axis` though that's a keyword argument rather that the name of the function. However the return type `std::span<const ssize_t> ` suggests to me to call it `axes`.

Reply:
I think we should keep the input keyword argument as `axis` for consistency with NumPy. But I see they do use both [axis and axes](https://numpy.org/doc/stable/glossary.html#term-axis) so I would rename this method and other places where we store multiple as `axes`.